src/shogun/labels/LabelEncoder.h

```diff
@@ -0,0 +1,146 @@
+		 * @return original encoding labels
+		 */
+		virtual std::shared_ptr<Labels>
+		inverse_transform(const std::shared_ptr<Labels>&) = 0;
+
+		/** Fit label encoder and return encoded labels.
+		 *
+		 * @param Target values.
+		 * @return Labels transformed to be normalized.
+		 */
+		virtual std::shared_ptr<Labels>
+		fit_transform(const std::shared_ptr<Labels>&) = 0;
+
+		virtual const char* get_name() const
+		{
+			return "LabelEncoder";
+		}
+
+	protected:
+		virtual bool check_is_contiguous(
+		    const SGVector<float64_t>& vec, const std::set<float64_t>& labels)
+		{
+			return false;
+		}
+		SGVector<float64_t> fit_impl(const SGVector<float64_t>& origin_vector)
+		{
+			is_fitted = true;
+			std::copy(
+			    origin_vector.begin(), origin_vector.end(),
+			    std::inserter(unique_labels, unique_labels.begin()));
+			if (check_is_contiguous(origin_vector, unique_labels))
+			{
+				is_fitted = false;
+			}
+			return SGVector<float64_t>(
+			    unique_labels.begin(), unique_labels.end());
+		}
+
+		SGVector<float64_t>
+		transform_impl(const SGVector<float64_t>& result_vector)
+		{
+			is_transformed = true;
+			if (!is_fitted && unique_labels.size())
+				return result_vector;
+			require(is_fitted, "Transform expect to be called after fit.");
+			SGVector<float64_t> converted(result_vector.vlen);
+			std::transform(
+			    result_vector.begin(), result_vector.end(), converted.begin(),
+			    [& unique_labels = unique_labels,
+			     &inverse_mapping = inverse_mapping](const auto& old_label) {
+				    auto new_label = std::distance(
+				        unique_labels.begin(), unique_labels.find(old_label));
+				    inverse_mapping[new_label] = old_label;
+				    return new_label;
+			    });
+			return converted;
+		}
+
+		SGVector<float64_t>
+		inverse_transform_impl(const SGVector<float64_t>& result_vector)
+		{
+			require(
+			    is_transformed,
+			    "Inverse transform expect to be called after transform.");
+			if (!is_fitted && unique_labels.size() && is_transformed)
+			{
+				return result_vector;
+			}
+			SGVector<float64_t> original_vector(result_vector.vlen);
+			std::transform(
+			    result_vector.begin(), result_vector.end(),
+			    original_vector.begin(),
+			    [& inverse_mapping = inverse_mapping](const auto& e) {
+				    return inverse_mapping[e];
+			    });
+			return original_vector;
+		}
+
+		bool can_convert_float_to_int(const SGVector<float64_t>& vec) const
+		{
+			SGVector<int32_t> converted(vec.vlen);
+			std::transform(
+			    vec.begin(), vec.end(), converted.begin(),
+			    [](auto&& e) { return static_cast<int32_t>(e); });
+			return std::equal(
+			    vec.begin(), vec.end(), converted.begin(),
+			    [&](auto&& e1, auto&& e2) {
+				    return Math::fequals(e1, static_cast<float64_t>(e2), eps);
+			    });
+		}
+
+		std::set<float64_t> unique_labels;
+		std::unordered_map<float64_t, float64_t> inverse_mapping;
+		const float64_t eps = std::numeric_limits<float64_t>::epsilon();
```

Comment:
```suggestion
		constexpr float64_t eps = std::numeric_limits<float64_t>::epsilon();
```

Reply:
when I changed `const` to `constexpr`, I got an error:  `non-static data member ‘eps’ declared ‘constexpr’`

Reply:
ah yes, it should be `static constexpr float64_t eps = std::numeric_limits<float64_t>::epsilon();`